Keycloak modules retry request on authentication error, support refresh token parameter

armkeh opened the pull request:
##### SUMMARY

Fixes #8857.

Wraps all requests to Keycloak in the Keycloak modules (`keycloak_authentication`, `keycloak_authz_authorization_scope`, `keycloak_authz_custom_policy`, etc.) with retry logic to make use of a new `refresh_token` module parameter.

This improves the user experience when using Keycloak modules with the `auth_token` parameter; previously if that token expired during playbook execution, subsequent tasks would fail. Now they "fall back" to using the `refresh_token`, or, if it is not provided or is expired itself, to using the `auth_username` and `auth_password`.

```python
    def _request(self, url, method, data=None):
        def make_request_ignoring_401():
            try:
                return open_url(url, method=method, data=data,
                                http_agent=self.http_agent, headers=self.restheaders,
                                timeout=self.connection_timeout,
                                validate_certs=self.validate_certs)
            except HTTPError as e:
                if e.code != 401:
                    raise e

            return None

        r = make_request_ignoring_401()
        if r is not None:
            return r

        # Authentication may have expired, re-authenticate with refresh token and retry
        refresh_token = self.module.params.get('refresh_token')
        if refresh_token is not None:
            token = _get_token_using_refresh_token(self.module.params)
            self.restheaders['Authorization'] = 'Bearer ' + token

        r = make_request_ignoring_401()
        if r is not None:
            return r

        # Retry once more with username and password
        auth_username = self.module.params.get('auth_username')
        auth_password = self.module.params.get('auth_password')
        if auth_username is not None and auth_password is not None:
            token = _get_token_using_credentials(self.module.params)
            self.restheaders['Authorization'] = 'Bearer ' + token

        return make_request_ignoring_401()
```

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
keycloak

russoz commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -142,6 +142,7 @@ def keycloak_argument_spec():
         validate_certs=dict(type='bool', default=True),
         connection_timeout=dict(type='int', default=10),
         token=dict(type='str', no_log=True),
+        refresh_token=dict(type='str', no_log=True),
```
You should add this to the doc_fragment as well ;-)

armkeh replied:
Done; thank you for pointing out the correct place to add it.

I marked it as added in 10.2.0, since I see that is the next expected version, and I'll update it if that is released before this is merged. I suggest we leave this conversation unresolved until checking that before merging.

russoz reviewed (commented):
hi @armkeh thanks for your contribution!

Couple of comments on the PR.

armkeh commented on plugins/modules/keycloak_authentication.py:
```diff
@@ -359,8 +359,11 @@ def main():
                            required_one_of=([['token', 'auth_realm', 'auth_username', 'auth_password']]),
-                           required_together=([['auth_realm', 'auth_username', 'auth_password']])
-                           )
+                           required_together=([['auth_username', 'auth_password']]),
```
Done. Ideally I wanted to enforce that `auth_realm` requires either `refresh_token` or both of `auth_username` and `auth_password`, but I don't see a means of doing so.

With this, to use `refresh_token`, all of `auth_realm`, `auth_username` and `auth_password` will be have to be provided.

russoz replied:
well, given that the logic of the code tries "normal auth" first, then "refresh token", then "normal auth" again, I think the old requirements still hold and the required_by you wrote last should do the trick for the refresh - it means the user cannot pass the refresh token alone and expect things to work.

russoz commented on plugins/modules/keycloak_authentication.py:
```diff
@@ -359,8 +359,10 @@ def main():
                            required_one_of=([['token', 'auth_realm', 'auth_username', 'auth_password']]),
-                           required_together=([['auth_realm', 'auth_username', 'auth_password']])
-                           )
+                           required_together=([['auth_realm', 'auth_username', 'auth_password']]),
+                           required_by={
+                               'refresh_token': 'auth_realm',
+                           })
```
A bit of a pain, but being helpful to those coming after us:
```suggestion
                           required_together=([['auth_realm', 'auth_username', 'auth_password']]),
                           required_by={'refresh_token': 'auth_realm'},
                           )
```
We get leaner PRs with this, like one for plugins/modules/keycloak_clientscope_type.py below

armkeh replied:
No problem, done.

russoz commented on plugins/doc_fragments/keycloak.py:
```diff
@@ -57,6 +57,12 @@ class ModuleDocFragment(object):
+    description:
+      - Authentication refresh token for Keycloak API.
+    type: str
+    version_added: 10.2.0
```
Sorry, 10.3.0 now. Other than this, LGTM.

russoz reviewed (approved):
LGTM

felixfontein commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -3191,18 +3040,17 @@ def delete_client_role_scope_from_realm(self, payload, clientid, realm="master")
         return self.get_client_role_scope_from_realm(clientid, realm)
 
-    def fail_open_url(self, e, msg, **kwargs):
+    def fail_request(self, e, msg, **kwargs):
```
The method `fail_open_url` must be kept (simply add a wrapper that calls `fail_request`) for backwards compatibility. The module utils is part of the public API of the collection.

felixfontein commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -280,6 +311,45 @@ def __init__(self, module, connection_header):
+            token = _get_token_using_credentials(self.module.params)
+            self.restheaders['Authorization'] = 'Bearer ' + token
+
+        return make_request_ignoring_401()
```
Shouldn't the last request be made without ignoring 401?

armkeh replied:
I've fixed this now; to make the code wrapping the re-requests more consistent, the last request still catches the 401, but it's re-raised afterwards (thanks @russoz for pointing out the failure to raise exceptions in earlier commits).

felixfontein reviewed (commented):
Thanks for your contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment). Thanks.

desand01 commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -280,6 +311,45 @@ def __init__(self, module, connection_header):
+
+        # Authentication may have expired, re-authenticate with refresh token and retry
+        refresh_token = self.module.params.get('refresh_token')
+        if refresh_token is not None:
+            token = _get_token_using_refresh_token(self.module.params)
+            self.restheaders['Authorization'] = 'Bearer ' + token
+
+        r = make_request_ignoring_401()
+        if r is not None:
+            return r
```
I think we should not make a new request if the access token has not been refreshed.
```suggestion
        if refresh_token is not None:
            token = _get_token_using_refresh_token(self.module.params)
            self.restheaders['Authorization'] = 'Bearer ' + token

            r = make_request_ignoring_401()
            if r is not None:
                return r
```

armkeh replied:
Thanks, good catch; moved the re-requests inside the if statements to avoid making them unnecessarily.

russoz commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -280,6 +311,46 @@ def __init__(self, module, connection_header):
         self.restheaders = connection_header
         self.http_agent = self.module.params.get('http_agent')
 
+    def _request(self, url, method, data=None):
+        def make_request_catching_401():
+            try:
+                return open_url(url, method=method, data=data,
+                                http_agent=self.http_agent, headers=self.restheaders,
+                                timeout=self.connection_timeout,
+                                validate_certs=self.validate_certs)
+            except HTTPError as e:
+                if e.code != 401:
+                    raise e
+                return e
+
+        r = make_request_catching_401()
+        if not isinstance(r, Exception):
+            return r
+
```
Little late in the game for that, I know, but I believe this would be leaner as:
```suggestion
    def _request(self, url, method, data=None):
        class Unauthorized(Exception):
            pass

        def make_request_catching_401():
            try:
                return open_url(url, method=method, data=data,
                                http_agent=self.http_agent, headers=self.restheaders,
                                timeout=self.connection_timeout,
                                validate_certs=self.validate_certs)
            except HTTPError as e:
                if e.code != 401:
                    raise e
                raise Unauthorized()

        try:
            return make_request_catching_401()
        exception Unauthorized:
            pass
        except Exception:
            raise
```
and so on

armkeh replied:
I like this, however we need to keep the original 401 exception in case no re-auth options are available and it needs to be re-raised. (Thanks for catching the mistake where it was being returned and not raised in your other comment.)

russoz commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -280,6 +311,46 @@ def __init__(self, module, connection_header):
+            token = _get_token_using_credentials(self.module.params)
+            self.restheaders['Authorization'] = 'Bearer ' + token
+
+            r = make_request_catching_401()
+
+        return r
+
```
This last call to `make_request_catching_401()` is not being verified. If, for whatever reason, the HTTP request inside that call returns a 401, that Exception object will be _returned_ (not raised) and no error handling will be performed.

armkeh replied:
Thanks, good catch; I've reworked the retry wrappers to make the flow a little clearer, removing early returns, and made sure to throw the exception if that's the final result.

russoz reviewed (commented):
Couple of comments more.

russoz commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -280,6 +316,53 @@ def __init__(self, module, connection_header):
+            except HTTPError as e:
+                if e.code != 401:
+                    raise e
+                return e
```
This still bugs me on principle - exceptions are meant to be raised not returned :-) - but let's not hold this PR back because of it. I'll make a note to myself to return to this later and see if I can make it any better.

russoz reviewed (commented):
one possible improvement and a comment, other than that LGTM

felixfontein commented:
LGTM as well.

armkeh commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -151,7 +152,79 @@ def camel(words):
+        raise KeycloakError(
+            'Could not obtain access token from %s' % auth_url)
+
+    return token
```
Done. Rather than exclude the `KeyError` catch, I made it's message clearer about the problem (response did not include `access_token`).

armkeh commented:
Thank you all for the feedback! I've addressed the comments (except @russoz's regarding the helper function returning an exception, for which a neat refactor isn't obvious), and added docstrings to the functions I've added or significantly reworked; so outside of any other feedback, I think this is ready to go.

One note, there are some integration tests failing for the Keycloak modules; I confirmed those same tests fail on the main branch currently. I'm planning to try and repair them in a separate PR.

felixfontein commented:
@armkeh thanks for your contribution!
@desand01 @russoz thanks for reviewing!

felixfontein commented on plugins/module_utils/identity/keycloak/keycloak.py:
```diff
@@ -1826,16 +1856,10 @@ def get_role_composites(self, rolerep, clientid=None, realm='master'):
-                headers=self.restheaders,
-                timeout=self.connection_timeout,
-                validate_certs=self.validate_certs).read()))
+            self._request_and_deserialize(composite_url, method='GET')
```
Here a `return` got lost.

felixfontein reviewed (commented):
I went through all `return` statements in the diff, this seems to be the only one missing.

Ref: #9678
